### PR TITLE
A compilable version of np.repeat.

### PR DESCRIPTION
**Pull request by jg8610**
A new keyword argument has been added to np.repeat, total_repeat_length, that can optionally be supplied to make np.repeat jit-able.

### CODE REVIEW

File: jax/numpy/lax_numpy.py [hunk 1]
```diff
@@ -2458,72 +2458,52 @@ def indices(dimensions, dtype=int32, sparse=False):
...
-  if size(repeats_raveled) == 1:
-    return _repeat_scalar(a, repeats_raveled.item(), axis)
-
-  if axis is None or isscalar(a):
-    a = ravel(a)
+
+@_wraps(np.repeat, lax_description=_TOTAL_REPEAT_LENGTH_DOC)
+def repeat(a, repeats, axis=None, *, total_repeat_length=None):
+  if axis is None:
     axis = 0
```
Comment (jakevdp) [inline jax/numpy/lax_numpy.py:2474]: I think when `axis is None` the input `a` should be raveled before proceeding; for example:
```python
>>> np.repeat(np.arange(4).reshape(2, 2), 2)
array([0, 0, 1, 1, 2, 2, 3, 3])
```

File: jax/numpy/lax_numpy.py [hunk 2]
```diff
@@ -2458,72 +2458,52 @@ def indices(dimensions, dtype=int32, sparse=False):
...
+  if ndim(a) == 0:
+    if ndim(repeats) == 0 or repeats.shape == (1,):
+      return full([total_repeat_length], a)
+    else:
+      raise ValueError('`repeat` with a scalar parameter `a` is only '
+      'implemented for scalar values of the parameter `repeats`.')
+
+  # Broadcast repeats. Takes a scalar e.g. `5` and broadcasts it to [5, ..., 5].
+  if ndim(repeats) == 0:
+    repeats = full([a.shape[axis]], repeats)
```
Comment (jakevdp) [inline jax/numpy/lax_numpy.py:2489]: You can use `np.broadcast_to` for this regardless of the dimension of `repeats`. That includes the error checking in the following lines.

File: jax/numpy/lax_numpy.py [hunk 3]
```diff
@@ -2458,72 +2458,52 @@ def indices(dimensions, dtype=int32, sparse=False):
...
+
+  if repeats.shape != a.shape[axis]:
+    raise ValueError(f"repeats shape {repeats.shape} does not match "
+                     f"the dimension on axis {a.shape[axis]}")
 
-  new_shape = list(a.shape)
-  new_shape[axis] = total
-  a_flattened = ravel(a)
+  exclusive_repeats = roll(repeats, shift=1)
+  exclusive_repeats = ops.index_update(exclusive_repeats, idx=0, y=0)
```
Comment (jakevdp) [inline jax/numpy/lax_numpy.py:2496]: A bit more concise: `exclusive_repeats = roll(repeats, 1).at[0].set(0)`

**Review by jakevdp — CHANGES_REQUESTED** [review]
Looks good! A couple small comments.

In addition, this PR should also add some test cases that cover the new functionality.

File: jax/numpy/lax_numpy.py [hunk 4]
```diff
@@ -2458,72 +2458,68 @@ def indices(dimensions, dtype=int32, sparse=False):
...
-    return _repeat_scalar(a, repeats_raveled.item(), axis)
-
-  if axis is None or isscalar(a):
+
+@_wraps(np.repeat, lax_description=_TOTAL_REPEAT_LENGTH_DOC)
+def repeat(a, repeats, axis=None, *, total_repeat_length=None):
+  # Ensure that `repeats` is an array.
+  repeats = array(repeats)
+  # Ravel repeats to remove any nested structure.
+  repeats = ravel(repeats)
```
Comment (jakevdp) [inline jax/numpy/lax_numpy.py:2476]: I would remove these comments – the code is pretty self-explanatory

File: jax/numpy/lax_numpy.py [hunk 5]
```diff
@@ -2458,72 +2458,68 @@ def indices(dimensions, dtype=int32, sparse=False):
...
-  if repeats_raveled.size != a.shape[axis]:
-    raise ValueError(f"repeats shape {repeats_raveled.shape} does not match "
-                     f"the dimension on axis {a.shape[axis]}")
+  # If repeats is on a zero sized axis, then return the array.
+  if a.shape[axis] == 0:
+    return a
+
+  # Special case when a is a scalar.
+  if ndim(a) == 0:
+    if ndim(repeats) == 0 or repeats.shape == (1,):
```
Comment (jakevdp) [inline jax/numpy/lax_numpy.py:2488]: `repeats` has already been raveled, so `ndim(repeats)` will never be `0`

File: jax/numpy/lax_numpy.py [hunk 6]
```diff
@@ -2458,72 +2458,68 @@ def indices(dimensions, dtype=int32, sparse=False):
...
+      if total_repeat_length is None:
+        return full(ravel(repeats), a)
```
Comment (jakevdp) [inline jax/numpy/lax_numpy.py:2490]: no need to ravel again.

File: jax/numpy/lax_numpy.py [hunk 7]
```diff
@@ -2458,72 +2458,68 @@ def indices(dimensions, dtype=int32, sparse=False):
...
+    else:
+      raise ValueError('`repeat` with a scalar parameter `a` is only '
+      'implemented for scalar values of the parameter `repeats`.')
 
-  # calculating the new shape
-  total = repeats_raveled.sum()
+  # Broadcasting the repeats if a scalar value.
+  if ndim(repeats) == 0 or repeats.shape[0] == 1:
+    # The ravel below ensures that a scalar repeat is always contained in [].
+    repeats = broadcast_to(ravel(repeats), [a.shape[axis]])
```
Comment (jakevdp) [inline jax/numpy/lax_numpy.py:2500]: I would use `broadcast_to` regardless of the number of dimensions or shape; this would simplify the code because it automatically takes care of the shape check that you do manually below, and raises an appropriate error.

File: jax/numpy/lax_numpy.py [hunk 8]
```diff
@@ -2458,72 +2458,68 @@ def indices(dimensions, dtype=int32, sparse=False):
...
-      ret = concatenate((ret, tile(a_splitted[i], (repeat,))))
+  # Modify repeats from e.g. [1,2,5] -> [0,1,2]
+  exclusive_repeats = roll(repeats, shift=1).at[0].set(0)
+  # Cumsum to get indices of new number in repeated tensor, e.g. [0, 1, 3]
+  scatter_indices = cumsum(exclusive_repeats)
+  # Scatter these onto a zero buffer, e.g. [1,1,0,1,0,0,0,0]
+  block_split_indicators = ops.index_update(
+      x=zeros([total_repeat_length], dtype=int32),
+      idx=scatter_indices,
+      y=ones_like(scatter_indices, dtype=int32))
```
Comment (jakevdp) [inline jax/numpy/lax_numpy.py:2518]: Rather than `y=ones_like(...)`, a scalar value `y=1` works here.

**Review by jakevdp — CHANGES_REQUESTED** [review]
Looking good! A few more minor comments, and we'll still need some tests of the new functionality before we can merge this. You can modify the existing test here: https://github.com/google/jax/blob/bf97e47929f22d04333141e3d93c02016aeb7738/tests/lax_numpy_test.py#L1299-L1316

File: jax/numpy/lax_numpy.py [hunk 9]
```diff
@@ -2458,72 +2458,63 @@ def indices(dimensions, dtype=int32, sparse=False):
...
+        return full(repeats, a)
+      else:
+        return full([total_repeat_length], a)
+    else:
+      raise ValueError('`repeat` with a scalar parameter `a` is only '
+      'implemented for scalar values of the parameter `repeats`.')
 
-  # calculating the new shape
-  total = repeats_raveled.sum()
+  repeats = broadcast_to(ravel(repeats), [a.shape[axis]])
```
Comment (jakevdp) [inline jax/numpy/lax_numpy.py:2495]: repeats has already been raveled; no need to do it again.

File: jax/numpy/lax_numpy.py [hunk 10]
```diff
@@ -2458,72 +2458,63 @@ def indices(dimensions, dtype=int32, sparse=False):
...
 
-  new_shape = list(a.shape)
-  new_shape[axis] = total
-  a_flattened = ravel(a)
+  # If total_repeat_length is not given, use a default.
+  if total_repeat_length is None:
+    total_repeat_length = sum(repeats)
```
Comment (jakevdp) [inline jax/numpy/lax_numpy.py:2499]: minor, but I would move this block to the top of the function, just after `repeats` is converted. I find it easier to reason about the function's behavior if default arguments are handled immediately.

As a side-benefit, this would let you simplify the conditional within the `ndim(a) == 0` case above.

File: jax/numpy/lax_numpy.py [hunk 11]
```diff
@@ -2458,72 +2458,63 @@ def indices(dimensions, dtype=int32, sparse=False):
...
-  # first break down raveled input array into list of chunks; each chunk is the
-  # unit of repeat. then tile the repeats to have same length as the list of
-  # chunks. finally repeat each unit x number of times according to the tiled
-  # repeat list.
-  chunks = _prod(a.shape[:axis+1])
-  a_splitted = split(a_flattened, chunks)
-  repeats_tiled = np.tile(repeats_raveled, chunks // len(repeats_raveled))
+  if repeats.shape[0] != a.shape[axis]:
+    raise ValueError(f"repeats shape {repeats.shape} does not match "
+                     f"the dimension on axis {a.shape[axis]}")
```
Comment (jakevdp) [inline jax/numpy/lax_numpy.py:2503]: This block can be removed: you've called `broadcast_to` on `repeats`, so either the shape will match, or an error will have already been raised.

File: tests/lax_numpy_test.py [hunk 12]
```diff
@@ -1309,11 +1309,22 @@ def testRepeat(self, axis, shape, dtype, repeats, rng_factory):
...
+    # Calculate expected size of the repeated axis.
+    # Broadcasting the repeats if a scalar value.
+    expected_repeats = repeats
+    expected_repeats = jnp.broadcast_to(jnp.ravel(expected_repeats),
+                                        [shape[axis]])
+    total_repeat_length = jnp.sum(expected_repeats)
+    if shape[axis] == 0:
+      total_repeat_length = 0
+    co_fun = lambda arg: jnp.repeat(arg, repeats=repeats, axis=axis,
+                                    total_repeat_length=total_repeat_length)
```
Comment (jakevdp) [inline tests/lax_numpy_test.py:1321]: We should not test multiple cases within one test case. Instead, please modify the parametrization to test arrays of repeats as well as scalars.

**Review by jakevdp — CHANGES_REQUESTED** [review]
Thanks! We're getting ever closer :grin:

File: tests/lax_numpy_test.py (hunk 12 above)
Comment (jg8610) [reply tests/lax_numpy_test.py:1321]: These were designed to check that, if total_repeat_length is set correctly, the results are the same as a without total_repeat_length and as numpy. What would the correct parametrization of that be? Perhaps what I'm testing is unclear  - should add some comments!

Comment (jakevdp) [reply tests/lax_numpy_test.py:1321]: Ah, I see. I was confused by the fact that you modified repeats in the process. Incidentally, I just noticed that becuase of the way Python variable scope works, this test no longer tests scalar inputs to `repeat` at all! The `lambda` functions in lines 1309 and 1311 access the `repeats` variable from this function scope, and by the time those functions are called below the scalar values have been replaced by an array.

As for how to parametrize this, I'd probably add a boolean parameter called something like `pass_total_repeat_length`, and pass the total repeat length only when it's true.

File: tests/lax_numpy_test.py [hunk 13]
```diff
@@ -1292,19 +1292,36 @@ def args_maker():
...
+        # Total size will be num_repeats X axis length.
+        total_repeat_length = jnp.sum(expected_repeats)
+    else:
+      total_repeat_length=None
 
+    jnp_fun = lambda arg: jnp.repeat(arg, repeats=repeats, axis=axis,
+                                     total_repeat_length=total_repeat_length)
     args_maker = lambda: [rng(shape, dtype)]
 
     self._CheckAgainstNumpy(np_fun, jnp_fun, args_maker)
```
Comment (jakevdp) [inline tests/lax_numpy_test.py:1327]: Last thing I would do here is add a ``self._CompileAndCheck`` in the case that `fixed_size` is True, because the function should be compilable in that case.

Comment (jg8610) [reply tests/lax_numpy_test.py:1327]: Funily enough, this is always compilable for the current test cases because they only have scalar repeats. I'll add some array repeats.

Comment (jg8610) [reply tests/lax_numpy_test.py:1327]: Oh I think I misunderstood - repeats is a constant in a closure in the current use case so that's while it is already compiling! My bad.

File: tests/lax_numpy_test.py [hunk 14]
```diff
@@ -1292,23 +1292,42 @@ def args_maker():
...
     rng = rng_factory(self.rng())
     np_fun = lambda arg: np.repeat(arg, repeats=repeats, axis=axis)
     np_fun = _promote_like_jnp(np_fun)
-    jnp_fun = lambda arg: jnp.repeat(arg, repeats=repeats, axis=axis)
-
-    args_maker = lambda: [rng(shape, dtype)]
+    if fixed_size:
+      # Calculate expected size of the repeated axis.
+      if axis is None:
+        axis = 0
```
Comment (jakevdp) [inline tests/lax_numpy_test.py:1311]: I think when `shape` is multi-dimensional, this leads to the wrong total repeat length, since the input is raveled when `axis is None`

**Review by jakevdp — CHANGES_REQUESTED** [review]
Looking good! A couple comments:
- It looks like the case of array-of-repeats with `total_repeat_size=None` is not currently covered in tests
- I think we should compare against numpy in all test cases, even when `total_repeat_size` is specified (unless we are deliberately testing the case when `total_repeat_size` does not match the numpy output)
- one comment below.

Maybe it would be cleaner to factor out a `_compute_total_repeat_length` helper that could be used both in the test and the implementation? That would avoid having such complicated logic within the test itself.

**Comment by jg8610** [comment]
Hey,

Thanks for your comments! 

So there are tests further down for multi-arrays, and I've added the compilation check to these when `repeats` is passed. Let me know if that covers your point.

In terms of the convenience function, in my understanding lax_numpy is exposed through the __init__ of jax/numpy. To access the private method from the test, I'd either have to expose it through __init__ or import lax_numpy directly. These seem a bit overkill, but happy to do so if you think it's worth it.

File: tests/lax_numpy_test.py [hunk 15]
```diff
@@ -1292,23 +1292,51 @@ def args_maker():
...
+      self._CheckAgainstNumpy(np_fun, clo_fun, clo_fun_args_maker)
+    else:
+      # Now repeats is in a closure, so a constant.
+      jnp_fun = lambda arg: jnp.repeat(arg, repeats=repeats, axis=axis)
+      args_maker = lambda: [rng(shape, dtype)]
+      self._CheckAgainstNumpy(np_fun, jnp_fun, args_maker)
+      self._CompileAndCheck(jnp_fun, args_maker)
 
-    self._CheckAgainstNumpy(np_fun, jnp_fun, args_maker)
-    self._CompileAndCheck(jnp_fun, args_maker)
```
Comment (jg8610) [inline tests/lax_numpy_test.py:1340]: These are the changes for multi dimensional arrays and compilation.

**Comment by jakevdp** [comment]
Since the "compute total repeats" operation is nontrivial and now duplicated in two places, I think at the very least it should be factored out into a helper function within the test file.

File: tests/lax_numpy_test.py [hunk 16]
```diff
@@ -1329,7 +1357,11 @@ def testUnique(self, shape, dtype, return_index, return_inverse, return_counts):
     jnp_fun = lambda x: jnp.unique(x, return_index, return_inverse, return_counts)
     self._CheckAgainstNumpy(np_fun, jnp_fun, args_maker)
 
-  def testIssue1233(self):
+  @parameterized.named_parameters(jtu.cases_from_list(
+      {"testcase_name": "_fixed_size={}".format(fixed_size),
+      "fixed_size": fixed_size}
+      for fixed_size in [True, False]))
+  def testIssue1233(self, fixed_size):
```
Comment (jakevdp) [inline tests/lax_numpy_test.py:1364]: Not related to your PR, but I'd rename this to something like `testNonScalarRepeat` or something more descriptive.

**Comment by jakevdp** [comment]
Looks great, thanks!

**Review by jakevdp — APPROVED** [review]
